File: Interviewbit/Binary Search/painters-partition-problem.java
public class Solution {
    int MOD=10000003;
	private boolean isPossible(long midT, int A,int []C){
		int count=1;
		long unit=0;

		for(int i=0;i<C.length;i++){
			if(C[i]>midT) return false;
			if(unit+C[i]>midT){
				count++;
				unit=C[i];
			}else{
				unit+=C[i];
			}
		}
		if(count<=A) return true;
		else return false;
	}

    public int paint(int A, int B, int[] C) {
    	int minT=Integer.MIN_VALUE;
    	int maxT=0;

    	for(int i=0;i<C.length;i++){
    		if(C[i]>minT) minT=C[i];
    		maxT=maxT+(C[i])%MOD;
    	}
        int ans=Integer.MAX_VALUE;
    	while(minT<=maxT){
    	    int midT=minT+(maxT-minT)/2;
    		if(isPossible(midT,A,C)==true){
    			ans=Math.min(ans,midT);
    			maxT=midT-1;
    		}else{
    			minT=midT+1;
    		}
    	}
    	return (int)((ans*(long)B)%MOD);
    }
}

File: Interviewbit/Binary Search/painters-partition-problem.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long
int const MOD=10000003;

bool isPossible(ll midT, int A, vector<int>&c){
    int count=1,unit=0;
    for(int i=0;i<c.size();i++){
        if(c[i]>midT) return false;
        if(unit+c[i]<=midT){
            unit+=c[i];
        }else{
            count++;
            unit=c[i];
        }
    }
    if(count<=A){
        return true;
    }else{
        return false;
    }
}

int Solution::paint(int A, int B, vector<int> &C) {
    ll minT=*max_element(C.begin(),C.end())%MOD;
    ll maxT=accumulate(C.begin(),C.end(),0)%MOD;
    ll ans=INT_MAX;
    while(minT<=maxT){
        ll midT=(minT+maxT)/2;
        if(isPossible(midT,A,C)){
            ans=min(ans,midT);
            maxT=midT-1;
        }else{
            minT=midT+1;
        }
    }
    return (ans*B)%MOD;
}
